### PR TITLE
Rewriting of the error handler

### PR DESCRIPTION
**Pull request by alallema**
# Pull Request

This PR tries to clean and organize the exceptions by following this guide meilisearch/integration-guides#19 

## PR checklist
- [x] Follow the Basic Error Handler rules to normalize `Exception`
- [x] Add tests for Exceptions

**Notes**
The duplication of code on `HTTP` methods in the `MeilisearchHttpResquest` and `CustomOKHttp` files will be treated in a PR linked to the client's refacto

### CODE REVIEW

File: src/main/java/com/meilisearch/sdk/MeiliSearchHttpRequest.java [hunk 1]
```diff
@@ -117,15 +114,15 @@ <T> String put(String api, T body) throws Exception, MeiliSearchApiException {
...
+     * @throws MeilisearchException if the response is an error
      */
-    String delete(String api) throws Exception, MeiliSearchApiException {
+    String delete(String api) throws MeilisearchException {
         HttpResponse<?> httpResponse =
                 this.client.put(
                         factory.create(HttpMethod.DELETE, api, Collections.emptyMap(), null));
         if (httpResponse.getStatusCode() >= 400) {
-            throw new MeiliSearchApiException(
+            System.out.println("STATUS CODE API EXCEPTION");
```
Comment (brunoocasali) [inline src/main/java/com/meilisearch/sdk/MeiliSearchHttpRequest.java:124]: 👀

File: src/main/java/com/meilisearch/sdk/exceptions/APIError.java [hunk 2]
```diff
@@ -21,4 +21,22 @@ public APIError(String message, String code, String type, String link) {
         this.type = type;
         this.link = link;
     }
+
+    @Override
+    public String toString() {
+        return "APIError: {"
```
Comment (brunoocasali) [inline src/main/java/com/meilisearch/sdk/exceptions/APIError.java:27]: I have some different feelings when I see this haha, but I don't know to improve it :|

File: src/main/java/com/meilisearch/sdk/http/ApacheHttpClient.java [hunk 3]
```diff
@@ -49,29 +48,28 @@ private HttpResponse<?> execute(HttpRequest<?> request) throws Exception {
                 getCallback(response));
         try {
             return response.thenApply(this::mapResponse).get();
-        } catch (CancellationException | ExecutionException e) {
-            // todo: throw dedicated exception
-            throw new Exception(e);
+        } catch (Exception e) {
+            throw new MeilisearchException(e);
```
Comment (brunoocasali) [inline src/main/java/com/meilisearch/sdk/http/ApacheHttpClient.java:52]: Maybe use a `MeilisearchApiException` here?

File: src/test/java/com/meilisearch/integration/ExceptionsTest.java [hunk 4]
```diff
@@ -21,16 +24,23 @@ static void cleanMeiliSearch() {
         cleanup();
     }
 
-    /** Test MeiliSearchApiException serialization and getters */
+    @Test
+    public void testMeilisearchCommunicationException() throws Exception {
+        String indexUid = "MeilisearchCommunicationException";
+        Client wrongClient = new Client(new Config("http://wrongurl:1234", "masterKey"));
+        assertThrows(MeilisearchCommunicationException.class, () -> wrongClient.getIndex(indexUid));
```
Comment (brunoocasali) [inline src/test/java/com/meilisearch/integration/ExceptionsTest.java:31]:
```suggestion

        assertThrows(MeilisearchCommunicationException.class, () -> wrongClient.getIndex(indexUid));
```

File: src/main/java/com/meilisearch/sdk/exceptions/MeilisearchCommunicationException.java [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+    }
+
+    public MeilisearchCommunicationException(Throwable cause) {
+        super(cause);
+        this.error = cause.toString();
+    }
+
+    @Override
+    public String toString() {
+        return "Meilisearch CommunicationException: {" + "error=" + this.error + '}';
```
Comment (brunoocasali) [inline src/main/java/com/meilisearch/sdk/exceptions/MeilisearchCommunicationException.java:32]: I missed a test that tests these methods (for each exception class) ;)

**Review by brunoocasali — COMMENTED** [review]
That was an awesome work @alallema 🎉 

I just left some comments for you ;)

File: src/main/java/com/meilisearch/sdk/http/ApacheHttpClient.java (hunk 3 above)
Comment (brunoocasali) [reply src/main/java/com/meilisearch/sdk/http/ApacheHttpClient.java:52]: I'm not sure about this comment anymore to be honest

File: src/main/java/com/meilisearch/sdk/exceptions/APIError.java (hunk 2 above)
Comment (alallema) [reply src/main/java/com/meilisearch/sdk/exceptions/APIError.java:27]: Me neither it was already like this ...

File: src/main/java/com/meilisearch/sdk/http/ApacheHttpClient.java (hunk 3 above)
Comment (alallema) [reply src/main/java/com/meilisearch/sdk/http/ApacheHttpClient.java:52]: The `ApacheHttpClient` will be removed so I admit that I didn't spend too much time on it

**Review by brunoocasali — APPROVED** [review]
🎉

**Pull request merged**